### PR TITLE
Make NotaryFlow idempotent

### PR DESCRIPTION
**Pull request by thschroeter**
Alternatively, we could make the underlying UniquenessProviders idempotent.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/flows/NotaryFlow.kt [hunk 1]
```diff
@@ -133,9 +133,18 @@ object NotaryFlow {
             try {
                 uniquenessProvider.commit(tx.inputs, tx.id, otherSide)
             } catch (e: UniquenessException) {
-                val conflictData = e.error.serialize()
-                val signedConflict = SignedData(conflictData, sign(conflictData.bytes))
-                throw NotaryException(NotaryError.Conflict(tx, signedConflict))
+                // Allow re-committing the transaction to make the NotaryFlow idempotent. Alternatively, we could make
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/flows/NotaryFlow.kt]: maybe you can write a method doc instead with a slightly more elaborate explanation that it commits the input states and only throws an exception if any of the states have been consumed by a different transaction

File: node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt [hunk 2]
```diff
@@ -85,28 +85,51 @@ class NotaryServiceTests {
...
         }
 
+        val firstAttempt = NotaryFlow.Client(stx)
+        val secondAttempt = NotaryFlow.Client(stx)
+        clientNode.services.startFlow(firstAttempt)
+        val future = clientNode.services.startFlow(secondAttempt)
+
+        net.runNetwork()
+
+        future.resultFuture.getOrThrow()
```
Comment (adagys) [inline node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt]: could add an assertion that both flows return the same signature

File: node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt [hunk 3]
```diff
@@ -85,28 +85,51 @@ class NotaryServiceTests {
...
+    }
+
+    @Test fun `should report conflict when inputs are reused accross transactions`() {
```
Comment (adagys) [inline node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt]: typo "across"

File: node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt [hunk 4]
```diff
@@ -85,28 +85,51 @@ class NotaryServiceTests {
         assertThat(ex.error).isInstanceOf(NotaryError.TimestampInvalid::class.java)
     }
 
-    @Test fun `should report conflict for a duplicate transaction`() {
+    @Test fun `should sign identical transaction multiple times (signing is idempotent)`() {
```
Comment (adagys) [inline node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt]: also worth adding a test where notarisation fails when we use the same state as multiple inputs in a transaction

Comment (mikehearn) [inline node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt]: Actually that sounds like it should be a validation rule - such a transaction should always be invalid!

**Comment by mikehearn** [comment]
Ball in your court @adagys

File: node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt (hunk 4 above)
Comment (adagys) [inline node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt]: Good point! @thschroeter can you move the duplicate input detection to `TransactionType.verify()`, and the test to `TransactionTests` (can do it as a separate PR)

File: core/src/main/kotlin/net/corda/flows/NotaryFlow.kt [hunk 5]
```diff
@@ -129,20 +131,49 @@ object NotaryFlow {
...
+                seenInputs += stateRef
+            }
+            if (conflicts.isNotEmpty()) {
+                throw notaryException(tx, UniquenessException(UniquenessProvider.Conflict(conflicts)))
+            }
+        }
+
+        /** A NotaryException is thrown if any of the states have been consumed by a different transaction. Note that
+         * this method does not throw an exception when input states are present multiple times within the transaction.
+         */
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/flows/NotaryFlow.kt]: minor: for multi-line comment blocks our convention is to add a line-break after `/**`

File: node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt (hunk 4 above)
Comment (mikehearn) [inline node/src/test/kotlin/net/corda/node/services/NotaryServiceTests.kt]: It's also fine to just drop a TODO in here and we'll do the move, if you'd prefer to do other things!

**Comment by thschroeter** [comment]
Added a TODO about moving the duplicate detection. I'll open a follow-up PR with the move (and ask for help if too many test break).

I'd like to start working on the bft-smart notary afterwards.

**Comment by thschroeter** [comment]
Thanks :)
